// @theintern/digdug/NullTunnel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var common_1 = require("@theintern/common");
var Tunnel_1 = require("./Tunnel");
var NullTunnel = (function (_super) {
    tslib_1.__extends(NullTunnel, _super);
    function NullTunnel(options) {
        return _super.call(this, Object.assign({
            auth: ''
        }, options || {})) || this;
    }
    Object.defineProperty(NullTunnel.prototype, "isDownloaded", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    NullTunnel.prototype.download = function () {
        return common_1.Task.resolve();
    };
    NullTunnel.prototype.start = function () {
        this._state = 'running';
        return common_1.Task.resolve();
    };
    NullTunnel.prototype.stop = function () {
        this._state = 'stopped';
        return Promise.resolve(0);
    };
    NullTunnel.prototype.sendJobState = function () {
        return common_1.Task.resolve();
    };
    return NullTunnel;
}(Tunnel_1.default));
exports.default = NullTunnel;
