// intern/lib/interfaces/bdd.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./tdd"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tdd_1 = require("./tdd");
    exports.describe = tdd_1.suite;
    exports.it = tdd_1.test;
    exports.before = tdd_1.before;
    exports.after = tdd_1.after;
    exports.beforeEach = tdd_1.beforeEach;
    exports.afterEach = tdd_1.afterEach;
    function getInterface(executor) {
        var _a = tdd_1.getInterface(executor), suite = _a.suite, test = _a.test, before = _a.before, after = _a.after, beforeEach = _a.beforeEach, afterEach = _a.afterEach;
        return {
            describe: suite,
            it: test,
            before: before,
            after: after,
            beforeEach: beforeEach,
            afterEach: afterEach
        };
    }
    exports.getInterface = getInterface;
});
